### ISSUE TITLE
GoDaddy Plugin fails removing acme challenge

### ISSUE THREAD

**Issue by davehope** [rmbolger/Posh-ACME#53]
The GoDaddy plugin fails to remove DNS records from my zone (hope.mx), failing as follows:

```
Invoke-RestMethod : {"code":"INACTIVE_DOMAIN","message":"The given domain is not eligible to have its nameservers changed"}
At C:\Program Files\WindowsPowerShell\Modules\Posh-ACME\2.4.0\DnsPlugins\GoDaddy.ps1:125 char:17
+     $response = Invoke-RestMethod -Uri "$apiRoot/$zone/records" `
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-RestMethod], WebException
    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeRestMethodCommand
```

This looks to be because the GoDaddy API doesn't support the removal of individual records. The plugin currently implements this by overwriting the entire zone.

My untested assumption is that this fails for my domain since it is locked with GoDaddy, meaning the NS records cannot be updated.

**Comment by davehope** [comment]
Since the GoDaddy API doesn't support the removal of individual records, the options as I see it are as follows:

1.  Continue with the current logic of replacing the entire zone (minus the record to be deleted) and update the readme. Possibly catch the exception and display something more friendly explaining why things failed. Those people with a locked domain will need to unlock their domains.

2. Investigate the potential of using the GoDaddy Domains API to unlock the domain, make the change and then re-lock the domain. This is risky as the subsequent lock call could fail, leaving the domain in a less-secure state. 

3. Replace the ACME challenge value with $null or similar

I'm happy to implement these, or another option, but would like to get a consensus on the preferred approach. My preference is the third option.

**Comment by rmbolger** [comment]
Man, that is a lame limitation on GoDaddy's part. I agree about option 3. It seems to be the least risky and will ultimately allow more people to use the plugin without hassle.

@Dule11 and @Rukas are the other users of the GoDaddy plugin I know of. You guys have an opinion?

From an implementation perspective, you might want to move the set-to-null logic into `Save-DnsTxtGoDaddy` to reduce the number of calls for certs with a lot of SANs. You can set a `$script:Blah` variable in the Remove function so that Save only runs the removal during the Remove batch and not the Add batch.  Check the Manual or AcmeDns plugins for an example.

**Comment by davehope** [comment]
It'd be great if either (or both!) @dule11 and @Rukas could chime in on whether their domains are locked or not,  and if they also get the error. Try as I might, I can't break this with GoDaddy's OTE environment

The GoDaddy API support team suggest it might be a timing issue, but I can't get it working by inserting significant delays between the creation of records and the removal. 

I'll wait to hear from others to confirm behaviour before implementing a fix.

**Comment by Dule011** [comment]
Before finding out about this module i was working on my own, rudimentary version and was OK with what you list as option 1. Yes it gets clunky as your zone grows, but in my testing, i have not found any issues (yet?). My domain is locked though, so i'm not sure about failures you see. I can take a look at what/how exactly am i removing records and will reply here on/after Monday (sorry, can't get to it any sooner).

**Comment by davehope** [comment]
Perfect, thanks.

Testing with my own domain, I still receive the INACTIVE_DOMAIN message when attempting to restore the zone contents even with the domain unlocked. So there might be more to it.

Here's a hacky script to test the add/remove functionality and report the lock status without the ACME bits:

https://gist.github.com/davehope/d902dae4680f7c81aa4fdff52921b203

If others aren't experiencing the same errors, it'd be useful to know the lock status and the tld of the domain.

**Comment by Dule011** [comment]
[GoDaddy add-update-remove F.zip](https://github.com/rmbolger/Posh-ACME/files/2134115/GoDaddy.add-update-remove.F.zip)
Attached is the function i was working on to handle CNAME and TXT records. so far still works (you can see type of tests i was doing at the bottom). Comparing to your code, i just can't see anything substantially different. It'd be odd if TLD had anything to do with this but in my test i was using .US. Hope this helps

**Comment by davehope** [comment]
Thanks @Dule011  , can you confirm whether your domain was locked during the tests?

**Comment by Dule011** [comment]
yes, it was locked

**Comment by davehope** [comment]
Currently working the issue through with the GoDaddy API team.

Waiting on some confirmation on the behaviour I'm seeing with a .mx domain to understand if there's anything TLD specific going on, or if the issue is limited to my domain for some reason.

**Comment by mithrandyr** [comment]
Here is a thought, instead of replacing the entire DNS zone -- only replace the record associated with the TXT type.  [GoDaddy Developer Documentation on Domains API](https://developer.godaddy.com/doc/endpoint/domains#/v1/recordReplaceType) -- Look for "PUT /v1/domains/{domain}/records/{type} -- I've successfully used this before to remove records.

**Comment by davehope** [comment]
Still working with the GoDaddy API team. It's a but frustrating as they've just managed to erase the contents of my zone.

If it doesn't progress I'll do what I mentioned earlier and just wipe the record contents.

**Comment by rmbolger** [comment]
Thanks for checking in. No worries. These things take time.

**Comment by davehope** [comment]
@rmbolger , Despite a back and forth over nearly two months, the GoDaddy API team have been unable to fix the error I reported initially ("The given domain is not eligible to have its nameservers changed"). This is despite having my domain and API Key details.

The GoDaddy API doesn't provide a method to remove individual records. The current method, as previously discussed, downloads the zone, removes the _acme-challenge then re-uploads the entire zone. This is, in my view, too risky. 

Therefore, our options are:

1. Do nothing, I might be one in a million. In which case I'll close the bug.
2. I can submit a PR that wipes out the content of Remove-DnsTxtGoDaddy, and just calls Add-DnsTxtGoDaddy with TxtValue set to a single space. In my testing, this works. 

I'd appreciate your thoughts before submitting the PR.

**Comment by rmbolger** [comment]
I haven't had a chance to sit down and look at this again yet. But my gut says option 2 is probably better but only in the case where there are no legit values left for a particular record. Like in the case where you're requesting a wildcard cert plus the root domain and there are 2 different values needed for the same TXT record `_acme-challenge.example.com`. `Remove-DnsTxtGoDaddy` is ultimately going to be called twice. And you wouldn't want to wipe out both values in a single call, only the value you're trying to remove. But if it's the only value left, then yeah, set it to an empty string.

**Comment by rmbolger** [comment]
Fixed with PR #71. Should go live with the next release.
